pkg/discovery/kubernetes/control-point-cache.go

```diff
@@ -0,0 +1,250 @@
+	"github.com/fluxninja/aperture/pkg/panichandler"
+	"google.golang.org/protobuf/proto"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/scale"
+)
+
+// A ControlPoint is identified by Group, Version, Kind, Namespace and Name.
+type ControlPoint struct {
+	Group     string
+	Version   string
+	Kind      string
+	Namespace string
+	Name      string
+}
+
+// ToProto converts a ControlPoint to a KubernetesControlPoint.
+func (cp *ControlPoint) ToProto() *controlpointcachev1.KubernetesControlPoint {
+	groupVersion := schema.GroupVersion{
+		Group:   cp.Group,
+		Version: cp.Version,
+	}
+
+	return &controlpointcachev1.KubernetesControlPoint{
+		ApiVersion: groupVersion.String(),
+		Kind:       cp.Kind,
+		Namespace:  cp.Namespace,
+		Name:       cp.Name,
+	}
+}
+
+// ControlPointFromSelector converts a policylangv1.KubernetesObjectSelector to a ControlPoint.
+func ControlPointFromSelector(k8sObjectSelector *policylangv1.KubernetesObjectSelector) (ControlPoint, error) {
+	// Convert Kubernetes APIVersion into Group and Version
+	groupVersion, parseErr := schema.ParseGroupVersion(k8sObjectSelector.ApiVersion)
+	if parseErr != nil {
+		log.Error().Err(parseErr).Msgf("Unable to parse APIVersion: %s", k8sObjectSelector.ApiVersion)
+		return ControlPoint{}, parseErr
+	}
+
+	return ControlPoint{
+		Group:     groupVersion.Group,
+		Version:   groupVersion.Version,
+		Kind:      k8sObjectSelector.Kind,
+		Namespace: k8sObjectSelector.Namespace,
+		Name:      k8sObjectSelector.Name,
+	}, nil
+}
+
+// AutoScaler provides an interface to invoke auto-scale.
+type AutoScaler interface {
+	Scale(ControlPoint, int32)
+}
+
+// ControlPointStore is the interface for Storing Kubernetes Control Points.
+type ControlPointStore interface {
+	Add(cp ControlPoint)
+	Update(cp ControlPoint)
+	Delete(cp ControlPoint)
+}
+
+// ControlPointCache is the interface for Reading or Watching Kubernetes Control Points.
+type ControlPointCache interface {
+	Keys() []ControlPoint
+	AddKeyNotifier(notifiers.KeyNotifier) error
+	RemoveKeyNotifier(notifiers.KeyNotifier) error
+}
+
+// controlPointCache is a cache of discovered Kubernetes control points and provides APIs to do CRUD on Scale type resources.
+type controlPointCache struct {
+	// RW controlPointsMutex
+	controlPointsMutex sync.RWMutex
+	scaleClient        scale.ScalesGetter
+	k8sClient          k8s.K8sClient
+	// Set of unique controlPoints
+	controlPoints map[ControlPoint]*controlPointState
+	trackers      notifiers.Trackers
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+// controlPointCache implements the ControlPointStore interface.
+var _ ControlPointStore = &controlPointCache{}
+
+// controlPointCache implements the ControlPointCache interface.
+var _ ControlPointCache = &controlPointCache{}
+
+// newControlPointCache returns a new ControlPointCache.
+func newControlPointCache(trackers notifiers.Trackers, k8sClient k8s.K8sClient) *controlPointCache {
```

Comment:
The return parameter can be the interface ControlPointCache

Reply:
Not really, because need to return an object that implements both `ControlPointCache` and `ControlPointStore` at the [caller](https://github.com/fluxninja/aperture/blob/tgill/k8s-control-point-discovery/pkg/discovery/kubernetes/provide.go#L70)